### DISCUSSION TITLE
The State of SSO (google,facebook,twitter,microsoft,github,discourse)

### DISCUSSION THREAD

**Discussion by Autr** [directus/directus#12443] [Archive]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

Hello, I thought I'd write up my findings on SSO (single-sign-on) for Directus on Directus latest `9.7.1` (28/03/22), which is something I'm keen to use since it simplifies user registration. 

Here is an overview followed by the full `.env` configuration:

_Last edited 20th Oct 2022_

### **Google** `openid`

✅ Works w/ setup via [cloud console](https://console.cloud.google.com/apis/credentials).

### **Facebook** `openid`

✅ Works w/ setup via [myapps](https://developers.facebook.com/apps/?show_reminder=true).

### **Discord** `oauth2`

✅ Works w/ setup via [apps](https://discord.com/developers/applications).
🛠 `AUTH_DISCORD_CLIENT_ID` must have "string:" prepended to it, else it is processed as an integer

### **Microsoft** `openid`

✅ Works w/ setup via [azure apps](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps)
🛠 For setup, everything is in hard-to-find [azure active directory > app registrations](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps).
🛠 If public-facing, it's also important to set `signInAudience=AzureADandPersonalMicrosoftAccount` inside *Manifest*. 
🛠 **Tenant ID** can be found inside [default directory > settings](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Properties).
🛠 **Client Secret** is also only valid w/ Azure for maximum 24 months (set a calendar alarm!). 

### **Twitter**  `oauth2`

❌ Configure via [apps](https://developer.twitter.com/en/apps), but not working yet.

* Initial [authorization url](https://developer.twitter.com/en/docs/authentication/api-reference/authorize) seems to be using OAUTH1.0 (unrecognised by Directus), whereas [access url](https://developer.twitter.com/en/docs/authentication/api-reference/token) is OAUTH2.0
* Twitter OAuth flow is unusually formatted. 

### **Github** `oauth2`

✅ Configure via [apps](https://github.com/settings/apps)
🛠 Only works if Github user email address is **public** if using email as `IDENTIFIER_KEY`

### **DiscourseHub** `openid`

✅ Configured via separate server running [distrust](https://github.com/Parkour-Vienna/distrust), which acts as middleman between _DiscourseConnect_ and _openid_, ie. `https://auth.your_domain.com`
🛠 Some more info on is in this [forum post](https://meta.discourse.org/t/using-discourse-as-an-identity-provider-sso-discourseconnect/32974)

# Environment (`.env`)

This is my current working environment (20th Oct 2022):

_NB: I've switched out the brand logos for each company, because I'm sick of looking at them, but they can easily be switched back_ 😅

```bash

# SSO ACCOUNTS
# ============

# Callback Format: {{url}}/auth/login/{{provider}}/callback
# Edit this to add to list of logins:

AUTH_PROVIDERS="string:scanlines,discord,github,microsoft,google,facebook"

# ✅ SCANLINES (openid) aka DiscourseHub
# ✅ DISCORD (oauth2)
# ✅ GITHUB (oauth2)
# ✅ GOOGLE (openid)
# ✅ MICROSOFT (openid)
# ✅ FACEBOOK (openid)
# ❌ TWITTER (oauth2)
# 🛠 SCANLINES, MICROSOFT, DISCORD & GITHUB have caveats (see details below)


# SCANLINES ✅ - openid
# ============

# 🛠 Install this on separate server: https://github.com/Parkour-Vienna/distrust
# ie. auth.discourse-instance.com
# Info A: https://meta.discourse.org/t/using-discourse-as-an-identity-provider-sso-discourseconnect/32974
# Info B: https://www.oauth.com/oauth2-servers/pkce/authorization-request/

AUTH_SCANLINES_DRIVER="openid"

AUTH_SCANLINES_CLIENT_ID="vidicon"
AUTH_SCANLINES_CLIENT_SECRET="XXXX"

AUTH_SCANLINES_AUTHORIZE_URL="https://auth.your_domain.com/oauth2/auth"
AUTH_SCANLINES_ISSUER_URL="https://auth.your_domain.com/oauth2/.well-known/openid-configuration"
AUTH_SCANLINES_PROFILE_URL="https://auth.your_domain.com/oauth2/userinfo"
AUTH_SCANLINES_SCOPE="openid"
AUTH_SCANLINES_IDENTIFIER_KEY="username"
AUTH_SCANLINES_EMAIL_KEY="email"
AUTH_SCANLINES_NAME_KEY="username"

AUTH_SCANLINES_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_SCANLINES_ICON="texture"
AUTH_SCANLINES_DEFAULT_ROLE_ID="XXXX"


# GOOGLE ✅ - openid
# =========

# Configure here: https://console.cloud.google.com/apis/credentials
# AUTH_GOOGLE_CLIENT_ID = XXXX with the App ID (Client ID)
# AUTH_GOOGLE_CLIENT_SECRET=  XXXX with the App Secret (Client Secret)

AUTH_GOOGLE_CLIENT_ID="XXXX.apps.googleusercontent.com" 
AUTH_GOOGLE_CLIENT_SECRET="XXXX" 
AUTH_GOOGLE_ISSUER_URL="https://accounts.google.com"

AUTH_GOOGLE_DRIVER="openid" 
AUTH_GOOGLE_IDENTIFIER_KEY="email"
AUTH_GOOGLE_EMAIL_KEY="email"
AUTH_GOOGLE_FIRST_NAME_KEY="given_name"
AUTH_GOOGLE_LAST_NAME_KEY="family_name"
AUTH_GOOGLE_ICON="image_search"
#"google"
AUTH_GOOGLE_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_GOOGLE_DEFAULT_ROLE_ID="XXXX"

# TWITTER ❌ - oauth2
# =========

# Configure here: https://developer.twitter.com/en/apps
# Doesn't work.

AUTH_TWITTER_DRIVER="oauth2"
AUTH_TWITTER_CLIENT_ID="XXXX"
AUTH_TWITTER_CLIENT_SECRET="XXXX"

AUTH_TWITTER_SCOPE="users.read"
AUTH_TWITTER_AUTHORIZE_URL="https://twitter.com/i/oauth2/authorize"
AUTH_TWITTER_ACCESS_URL="https://api.twitter.com/2/oauth2/token"
AUTH_TWITTER_PROFILE_URL="https://api.twitter.com/2/users/me"

AUTH_TWITTER_IDENTIFIER_KEY="data.username"
AUTH_TWITTER_EMAIL_KEY="data.username"
AUTH_TWITTER_FIRST_NAME_KEY="data.username"
AUTH_TWITTER_LAST_NAME_KEY="data.username"

AUTH_TWITTER_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_TWITTER_ICON="sports_kabaddi" 
# "twitter"
AUTH_TWITTER_DEFAULT_ROLE_ID="XXXX"


# MICROSOFT ✅
# ============

# Configure here: https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps
# 🛠 Make sure to set signInAudience=AzureADandPersonalMicrosoftAccount, in Manifest

AUTH_MICROSOFT_DRIVER="openid"

AUTH_MICROSOFT_CLIENT_ID="XXXX"
AUTH_MICROSOFT_CLIENT_SECRET="XXXX"
AUTH_MICROSOFT_AUTHORIZE_URL="https://login.microsoftonline.com/XXXX/oauth2/v2.0/authorize"
AUTH_MICROSOFT_ISSUER_URL="https://login.microsoftonline.com/XXXX/v2.0/.well-known/openid-configuration"


AUTH_MICROSOFT_SCOPE="openid profile email"

AUTH_MICROSOFT_IDENTIFIER_KEY="email"
AUTH_MICROSOFT_EMAIL_KEY="email"
AUTH_MICROSOFT_FIRST_NAME="firstname"
AUTH_MICROSOFT_LAST_NAME="lastname"

AUTH_MICROSOFT_ICON="point_of_sale"
AUTH_MICROSOFT_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_MICROSOFT_DEFAULT_ROLE_ID="XXXX"

# DISCORD ✅ - oauth2
# ==========

# Configure here: https://discord.com/developers/applications
# 🛠 AUTH_DISCORD_CLIENT_ID must have "string:" prepended to it, else it is processed as an integer

AUTH_DISCORD_CLIENT_ID="string:XXXX"
AUTH_DISCORD_CLIENT_SECRET="XXXX"

AUTH_DISCORD_AUTHORIZE_URL="https://discord.com/api/oauth2/authorize"
AUTH_DISCORD_ACCESS_URL="https://discord.com/api/oauth2/token"
AUTH_DISCORD_PROFILE_URL="https://discord.com/api/users/@me"

AUTH_DISCORD_SCOPE="identify email"

AUTH_DISCORD_IDENTIFIER_KEY="id"
AUTH_DISCORD_EMAIL_KEY="email"
AUTH_DISCORD_FIRST_NAME_KEY="username"

AUTH_DISCORD_DRIVER="oauth2"

AUTH_DISCORD_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_DISCORD_ICON="sports_esports"
# "discord"
AUTH_DISCORD_DEFAULT_ROLE_ID="XXXX"


# FACEBOOK ✅ - oauth2
# ===========

# Configure: https://developers.facebook.com/apps

AUTH_FACEBOOK_DRIVER="openid"
AUTH_FACEBOOK_CLIENT_ID="XXXX" 
AUTH_FACEBOOK_CLIENT_SECRET="XXXX" 

AUTH_FACEBOOK_SCOPE="email"
AUTH_FACEBOOK_AUTHORIZE_URL="https://www.facebook.com/dialog/oauth"
AUTH_FACEBOOK_ACCESS_URL="https://graph.facebook.com/oauth/access_token"
AUTH_FACEBOOK_PROFILE_URL="https://graph.facebook.com/me?fields=id,name,email,first_name,last_name"

AUTH_FACEBOOK_IDENTIFIER_KEY="id"
AUTH_FACEBOOK_EMAIL_KEY="email"
AUTH_FACEBOOK_FIRST_NAME_KEY="first_name"
AUTH_FACEBOOK_LAST_NAME_KEY="last_name"

AUTH_FACEBOOK_ICON="mood_bad" 
# "facebook"
AUTH_FACEBOOK_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_FACEBOOK_DRIVER="oauth2"
AUTH_FACEBOOK_DEFAULT_ROLE_ID="XXXX"

# GITHUB ✅ - oauth2
# =========

# Configure here: https://github.com/settings/apps
# Return: https://admin.vidicon.org/auth/login/github/callback
# 🛠 EMAIL MUST BE PUBLIC TO WORK: https://github.com/settings/profile

AUTH_GITHUB_CLIENT_ID="XXXX"
AUTH_GITHUB_CLIENT_SECRET="XXXX"

AUTH_GITHUB_SCOPE="email"
AUTH_GITHUB_AUTHORIZE_URL="https://github.com/login/oauth/authorize"
AUTH_GITHUB_ACCESS_URL="https://github.com/login/oauth/access_token"
AUTH_GITHUB_PROFILE_URL="https://api.github.com/user"

AUTH_GITHUB_ICON="code"
# "github"
AUTH_GITHUB_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_GITHUB_DRIVER="oauth2"
AUTH_GITHUB_DEFAULT_ROLE_ID="XXXX"

```

Ubuntu Droplet (Digital Ocean)
Directus 9.11.1 

_Last edited 20th Oct 2022_

**Comment by Autr** [comment]
PS. this is for a community video art festival

**Comment by aidenfoxx** [comment]
Great initiative on doing the research!

OAuth/OpenID is a nightmare since it can vary wildly between provider implementations, but we've tried to build the system as generically as possible to support as much as possible. Often provider quirks can be worked around by providing the right config, either through the  `AUTH_<PROVIDER>_PARAMS_` or `AUTH_<PROVIDER>_CLIENT_` to customize the authorization parameters and client configurations respectively.

Implementations I've personally verified are: facebook, github, discord, okta, auth0, keycloak, twitch and google.

Github requires the authenticating user to have their email as public in their settings in order for it to be picked up when authenticating. There might be a way around it by providing the proper grants or scope, but I never found it. That should probably be documented somewhere! 😅 

Microsoft has been confirmed working by a few users in discussions. I just pulled this example (you can exclude `AUTH_MICROSOFT_EMAIL_KEY`) https://github.com/directus/directus/discussions/11316#discussioncomment-2101655

Twitter is a lost cause, however. Their OAuth 2 implementation is so bespoke that it requires a targeted implementation (unless it's changed during the beta).

**Reply by SamsonChungStackAdapt** [reply]
Hi, could you please share your Okta implementation?

Thanks!

**Reply by aidenfoxx** [reply]
I don't have it written down but it should be fairly standard.
```
AUTH_OKTA_DRIVER="openid"
AUTH_OKTA_CLIENT_ID="<your_okta_id>"
AUTH_OKTA_CLIENT_SECRET= "<your_okta_secret_key>"
AUTH_OKTA_ISSUER_URL="https://<your_okta_domain>"
AUTH_OKTA_IDENTIFIER_KEY="email"
```

**Comment by aidenfoxx** [comment]
Actually it looks like Twitter has changed things. You can try `AUTH_TWITTER_AUTHORIZE_URL="https://twitter.com/i/oauth2/authorize"` and it may work. 🤞

**Comment by Autr** [comment]
Thanks @aidenfoxx ! 

Public Github email fixed it so I’ve updated the original post. 

`AUTH_TWITTER_AUTHORIZE_URL="https://twitter.com/i/oauth2/authorize"` no such luck but perhaps more `AUTH_TWITTER_PARAMS_` experimentation will do it.

Recreating [#11316](https://github.com/directus/directus/discussions/11316#discussioncomment-2101655) keys for `MICROSOFT` still same error, so perhaps some config needed with Azure settings.

Would you be able to share your Discord, Twitch configs? It would be great to document more examples. 💪

**Comment by Autr** [comment]
Here's a hack for Github SSO button:

```css
.sso-link:nth-child(5) .sso-title {
    display: flex!important;
    flex-direction: column!important;
    justify-content: center!important;
    align-items: flex-start!important;
}
.sso-link:nth-child(5) .sso-title:after {
    content:  "Public email address only";
    font-style:  italic!important;
    font-size: 14px!important;
}
```
<img width="282" alt="Screenshot 2022-03-28 at 20 08 29" src="https://user-images.githubusercontent.com/56603487/160459989-012206be-f1a2-494c-a03e-eb2b4d9dea94.png">

**Comment by aidenfoxx** [comment]
Sure thing!

Twitch:
```
AUTH_TWITCH_DRIVER="openid"
AUTH_TWITCH_ISSUER_URL="https://id.twitch.tv/oauth2/.well-known/openid-configuration"
AUTH_TWITCH_CLIENT_ID="123"
AUTH_TWITCH_CLIENT_SECRET="123"
AUTH_TWITCH_SCOPE="openid user:read:email"
AUTH_TWITCH_PARAMS__CLAIMS="string:{"id_token":{"email":null,"email_verified":null}}"
```

Discord:
```
AUTH_DISCORD_DRIVER="oauth2"
AUTH_DISCORD_AUTHORIZE_URL="https://discord.com/api/oauth2/authorize"
AUTH_DISCORD_ACCESS_URL="https://discord.com/api/oauth2/token"
AUTH_DISCORD_PROFILE_URL="https://discord.com/api/users/@me"
AUTH_DISCORD_CLIENT_ID="string:123"
AUTH_DISCORD_CLIENT_SECRET="123"
```

If I remember correctly, Discord uses numeric client ids, so you need to specify `string:` to keep things happy.

**Comment by Autr** [comment]
Awesome :)

~~I've discovered a funny bug in Directus for Discord. If the _Client ID_ is an integer then it causes Directus to crash.`AUTH_DISCORD_CLIENT_ID="123456"` or `AUTH_DISCORD_CLIENT_ID=123456`: But if I set `AUTH_DISCORD_CLIENT_ID` with string chars I don't get the bug.~~

Just noticed the flag **string** in the variable.

**Comment by aidenfoxx** [comment]
Looking further into the docs, the correct Twitter profile URL would be `AUTH_TWITTER_PROFILE_URL="https://api.twitter.com/2/users/me"`.

I think you also need additional scopes `AUTH_TWITTER_SCOPE="email users.read"`, and they also don't provide email, so you need to use the username as the identifier `AUTH_TWITTER_IDENTIFIER_KEY="username"`

**Comment by Autr** [comment]
Yep, before moving back onto Twitter I think this is the same issue as with Discord, where as with Github the page refreshes without creating the new user - presumably because there is no email to use as a key?

Can you explain how this works? ie. How does the OAuth2 flow get some field to use as a unique key / email for Directus?

~~This is my config for Discord:~~

_EDIT: Discord now work with this config:_

```
# DISCORD ✅
# ==========

# WHERE: https://discord.com/developers/applications
# CALLBACK: https://dev.vidicon.org/auth/login/discord/callback
# Discord: add callback to "CUSTOM URL"
# CLIENT_ID must have "string:" prepended to it (integer)

AUTH_DISCORD_CLIENT_ID="string:XXXX"
AUTH_DISCORD_CLIENT_SECRET="XXXX"

AUTH_DISCORD_AUTHORIZE_URL="https://discord.com/api/oauth2/authorize"
AUTH_DISCORD_ACCESS_URL="https://discord.com/api/oauth2/token"
AUTH_DISCORD_PROFILE_URL="https://discord.com/api/users/@me"

AUTH_DISCORD_SCOPE="identify"
AUTH_DISCORD_IDENTIFIER_KEY="username"
AUTH_DISCORD_DRIVER="oauth2"
AUTH_DISCORD_ALLOW_PUBLIC_REGISTRATION="true" 
AUTH_DISCORD_ICON="discord"
AUTH_DISCORD_DEFAULT_ROLE_ID="XXXX"
```

_Checking to see if this isn't the same for Github..._

**Comment by aidenfoxx** [comment]
The `IDENTIFIER_KEY` and `EMAIL_KEY` specify the keys to extract user data when reading the profile response from the SSO provider. The `IDENTIFIER_KEY` specifies the profile key that links the SSO user to a Directus user (as defined in "External Identifier" in the admin). If no `IDENTIFIER_KEY` is specified in the config we fallback to `EMAIL_KEY` as the identifier.

So if Twitter returns the following profile for the authenticating user:
```
{
  "id": "2244994945",
  ...
  "username": "SomeUser"
}
```
Then authenticating with `AUTH_TWITTER_IDENTIFIER_KEY="username"`, we'd expect a Directus user with the "Provider" value "Twitter" and the "External Identifier" value "SomeUser".

Unfortunately I see the issue with Twitter now. Their actual profile response is in a sub-object, and we have no way to query sub-objects in the current implementation:

```
{
  "data": {
    "id": "2244994945",
    "name": "TwitterDev",
    "username": "Twitter Dev"
  }
}
```

**Reply by StobyWan** [reply]
is this something that can be resolved in a new release or just not possible?

**Reply by aidenfoxx** [reply]
Very easily solved. We just need to flatten the twitter response so we can access the data via dot indexes like `AUTH_TWITTER_IDENTIFIER_KEY="data.username"`. I'll make a PR for it today.

**Reply by Autr** [reply]
`”data?.[0]?.data?.properties?.username"` 🤪

This would be a great! 

What do you think to also adding a `AUTH_[PROVIDER]_DEBUG="true|false"` for logging out req and responses on each endpoint? Could be a security risk, so maybe something like: `AUTH_[PROVIDER]_LOG="false|verbose|masked"`, where _"masked"_ shows keys but not values.

**Reply by aidenfoxx** [reply]
Twitter is confirmed working with the this PR https://github.com/directus/directus/pull/12457

**Reply by aidenfoxx** [reply]
@autr As for your request logging suggestion, the important requests are all handled inside `openid-client` so we couldn't log them even if we wanted to.

**Comment by Autr** [comment]
OK thank you, that makes perfect sense. 

What is the best way to see the response via an npm instance of Directus (ie. `directus start`)?

Twitter and Microsoft I'm going to park for now, but will try with others.

**Comment by aidenfoxx** [comment]
Well you can get more info about what's going wrong by enabling trace logging. Other than that I'd probably manually modify the "node_modules/directus/dist/auth/drivers/oauth2.js" and log `userInfo` in the `getUserID` function.

**Comment by azrikahar** [comment]
Moved this to discussions as it seems somewhat more exploratory in nature, yet understandably can be deemed as bugs for the ones that are (yet) to be working in the current results, but I digress 😄

---

@autr thank you very much for such an in depth verification on all the auth providers! This is quite literally a gold mine of information. cc @erondpowell so that we can enhance the docs with these findings/confirmations later on.

Regarding the Discord client ID being cast as number, it is a known issue tracked in #9521, and `string:` prefix is indeed the current workaround as Aiden mentioned 👍

---

@aidenfoxx just wanted to take this opportunity to thank you again for all the responses in any auth related issues/discussions. Sincerely appreciate it!

**Comment by Autr** [comment]
Hi all, I just edited the original post with my current working config - all SSOs are working except Twitter, and DiscourseHub is working via an intermediate server. 

```
# ✅ SCANLINES (openid) aka DiscourseHub
# ✅ DISCORD (oauth2)
# ✅ GITHUB (oauth2)
# ✅ GOOGLE (openid)
# ✅ MICROSOFT (openid)
# ✅ FACEBOOK (openid)
# ❌ TWITTER (oauth2)
# 🛠 SCANLINES, MICROSOFT, DISCORD & GITHUB have caveats
```

Scroll up to original post for details. Cheers!

**Reply by aidenfoxx** [reply]
Twitter should work (with caveats). Try the config here https://docs.directus.io/self-hosted/sso-examples.html#twitter

EDIT: Your config is super close to the one in the docs, so something is weird. I have personally tested that one in the past

**Reply by aidenfoxx** [reply]
Be sure you're using the OAuth 2.0 tokens and not the standard ones. Here is my working config in Twitter:

![settings](https://user-images.githubusercontent.com/5548518/197016978-1d857167-dfd7-4060-ae11-0cfbeac0cd74.png)

**Comment by meteoro** [comment]
Hi,
tested and working for Nextcloud 29.x

      AUTH_NEXTCLOUD_DRIVER: "oauth2"
      AUTH_NEXTCLOUD_SCOPE: "ocs.data.email ocs.data.displayname ocs.data.id"
      AUTH_NEXTCLOUD_CLIENT_ID: "<ID>"
      AUTH_NEXTCLOUD_CLIENT_SECRET: "<SECRET>"
      AUTH_NEXTCLOUD_AUTHORIZE_URL: "https://<DOMAIN>/apps/oauth2/authorize"
      AUTH_NEXTCLOUD_ACCESS_URL: "https://<DOMAIN>/apps/oauth2/api/v1/token"
      AUTH_NEXTCLOUD_PROFILE_URL: "https://<DOMAIN>/ocs/v2.php/cloud/user?format=json"
      AUTH_NEXTCLOUD_ALLOW_PUBLIC_REGISTRATION: "true"
      AUTH_NEXTCLOUD_IDENTIFIER_KEY: "ocs.data.id"
      AUTH_NEXTCLOUD_EMAIL_KEY: "ocs.data.email"
      AUTH_NEXTCLOUD_FIRST_NAME_KEY: "ocs.data.displayname"
      AUTH_NEXTCLOUD_DEFAULT_ROLE_ID: "<ROLE_ID>"
      AUTH_NEXTCLOUD_ICON: "cloud"
